### PR TITLE
refactor and test: ability for L2TargetDispenser to migrate

### PR DESCRIPTION
**Pull request by kupermind**
- Ability for L2TargetDispenser to migrate;
- Adding tests.

### CODE REVIEW

File: contracts/staking/DefaultTargetDispenserL2.sol [hunk 1]
```diff
@@ -393,6 +405,49 @@ abstract contract DefaultTargetDispenserL2 is IBridgeErrors {
         _locked = 1;
     }
 
+    /// @dev Migrates funds to a new specified L2 target dispenser contract address.
+    /// @notice The contract must be paused to prevent other interactions.
+    /// @notice Owner will be zeroed and the contract becomes paused for good.
+    function migrate(address newL2TargetDispenser) external {
```
Comment (kupermind) [inline contracts/staking/DefaultTargetDispenserL2.sol:411]: This function migrates all the funds to the new L2 target dispenser, and freezes all the contract activity.

Comment (kupermind) [reply contracts/staking/DefaultTargetDispenserL2.sol:411]: @DavidMinarsch @mariapiamo this has been implemented to address a preliminary discussion stating that if anything goes wrong on L1, the funds are stuck in this contract forever. This proposed implementation allows to deploy another L2 contract and migrate all the funds there.

File: contracts/staking/GnosisTargetDispenserL2.sol [hunk 2]
```diff
@@ -53,12 +55,25 @@ contract GnosisTargetDispenserL2 is DefaultTargetDispenserL2 {
...
+            revert IncorrectDataLength(BRIDGE_PAYLOAD_LENGTH, bridgePayload.length);
+        }
+
+        // Get the gas limit from the bridge payload
+        uint256 gasLimitMessage = abi.decode(bridgePayload, (uint256));
+
+        // Check the gas limit value
+        if (gasLimitMessage < GAS_LIMIT) {
+            gasLimitMessage = GAS_LIMIT;
+        }
```
Comment (kupermind) [inline contracts/staking/GnosisTargetDispenserL2.sol:70]: @DavidMinarsch @mariapiamo re: L2 to L1 message passing of a single amount variable. The bridged payload is checked for the gas limit, which is always bound by the number sufficient for the message delivery from L2 to L1. The value passed can always be bigger than that limit.

File: contracts/TokenomicsConstants.sol [hunk 3]
```diff
@@ -34,15 +34,15 @@ abstract contract TokenomicsConstants {
         // For the first 10 years the supply caps are pre-defined
         if (numYears < 10) {
             uint96[10] memory supplyCaps = [
-                529_659_000_00e16,
```
Comment (kupermind) [inline contracts/TokenomicsConstants.sol:37]: @mariapiamo please verify the numbers 🥇

Comment (mariapiamo) [reply contracts/TokenomicsConstants.sol:37]: lgtm

**Review by mariapiamo — APPROVED** [review]

File: contracts/staking/DefaultTargetDispenserL2.sol (hunk 1 above)
Comment (mariapiamo) [reply contracts/staking/DefaultTargetDispenserL2.sol]: Good! Even if the DAO would always be able to "drain" eventual withheld tokens with new staking contract, this removes completely the need of manual interaction for "fund recovery".

**Pull request merged**